From 0a2b0cda2fd72726809c9528e78bd5aa034c1b16 Mon Sep 17 00:00:00 2001
From: Tyler Gunn <tgunn@google.com>
Date: Wed, 14 Feb 2018 14:19:42 -0800
Subject: [PATCH] Enhanced permission checks for TelephonyManager#endCall()
 API.

The existing permission checks require CALL permission in order to end
an ongoing call or reject an incoming call.  The Telecom equivalent of
this API requires MODIFY_PHONE_STATE.  Since this API IS in fact a
system
API, modifying the permission requirements to require the same
MODIFY_PHONE_STATE permission.

Test: Manual testing per bug reproduction steps.
Bug: 67862398
Change-Id: I8f15f6afd00466f6b861079a83e9fa03b962f63f
(cherry picked from commit 6707357c6e198287046e76c1d6fc5edf1fcb3fee)
(cherry picked from commit ee1901e859ed4624723d18497b70c59b0b658747)
---
 src/com/android/phone/PhoneInterfaceManager.java | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index 31efd0f..c57c9d4 100644
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -22,6 +22,7 @@ import android.bluetooth.IBluetoothHeadsetPhone;
 import android.content.ActivityNotFoundException;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.PackageManager;
 import android.net.ConnectivityManager;
 import android.net.Uri;
 import android.os.AsyncResult;
@@ -39,6 +40,7 @@ import android.telephony.NeighboringCellInfo;
 import android.telephony.CellInfo;
 import android.telephony.ServiceState;
 import android.text.TextUtils;
+import android.util.EventLog;
 import android.util.Log;
 
 import com.android.internal.telephony.CallManager;
@@ -357,7 +359,12 @@ public class PhoneInterfaceManager extends ITelephony.Stub implements CallModele
      * @return true is a call was ended
      */
     public boolean endCall() {
-        enforceCallPermission();
+        if (mApp.checkCallingOrSelfPermission(Manifest.permission.MODIFY_PHONE_STATE)
+                != PackageManager.PERMISSION_GRANTED) {
+            Log.i(LOG_TAG, "endCall: called without modify phone state.");
+            EventLog.writeEvent(0x534e4554, "67862398", -1, "");
+            throw new SecurityException("MODIFY_PHONE_STATE permission required.");
+        }
         return (Boolean) sendRequest(CMD_END_CALL, null);
     }
 
-- 
2.7.4

